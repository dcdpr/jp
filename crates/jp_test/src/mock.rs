use std::{
    env, fs,
    path::{Path, PathBuf},
};

use httpmock::{MockServer, RecordingRuleBuilder};
use saphyr::{LoadableYamlNode, Yaml, YamlEmitter};

/// A recorder for HTTP requests.
pub struct Vcr {
    forward_to: String,
    fixtures: PathBuf,
    recording: bool,
}

impl Vcr {
    /// Create a new recorder.
    pub fn new(forward_to: impl Into<String>, fixtures: impl Into<PathBuf>) -> Self {
        Self {
            forward_to: forward_to.into(),
            fixtures: fixtures.into(),
            recording: env::var("RECORD").is_ok(),
        }
    }

    /// Enable recording for the current test.
    pub fn record(&mut self) {
        self.recording = true;
    }

    /// Enable playback for the current test.
    pub fn playback(&mut self) {
        self.recording = false;
    }

    /// Set recording mode for the current test.
    pub fn set_recording(&mut self, recording: bool) {
        self.recording = recording;
    }

    /// Record a test cassette.
    pub async fn cassette<R, F, Fut, O>(
        &self,
        name: &str,
        rule_builder: R,
        test_fn: F,
    ) -> Result<(), Box<dyn std::error::Error>>
    where
        R: FnOnce(RecordingRuleBuilder),
        F: FnOnce(bool, String) -> Fut,
        Fut: Future<Output = O>,
        O: std::fmt::Debug,
    {
        let fixture = self.fixtures.join(format!("{name}.yml"));
        let server = MockServer::start_async().await;

        if self.recording {
            server
                .forward_to_async(&self.forward_to, |rule| {
                    rule.filter(|when| {
                        when.any_request();
                    });
                })
                .await;

            let recording = server.record_async(rule_builder).await;

            let out = test_fn(true, server.base_url()).await;
            self.verify(name, out);

            let temp_path = recording.save_to_async(&self.fixtures, name).await?;

            if temp_path != fixture && temp_path.exists() {
                fs::rename(temp_path, &fixture)?;
            }

            modify_fixture(&fixture)?;
        } else if fixture.exists() {
            server.playback_async(&fixture).await;
            let out = test_fn(false, server.base_url()).await;
            self.verify(name, out);
        } else {
            return Err(format!("Recording not found at {}", fixture.display()).into());
        }

        Ok(())
    }

    fn verify<T: std::fmt::Debug>(&self, name: &str, expr: T) {
        insta::with_settings!({ snapshot_path => &self.fixtures, prepend_module_to_snapshot => false }, {
            insta::assert_debug_snapshot!(name, expr);
        });
    }
}

fn modify_fixture(fixture: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let contents = fs::read_to_string(fixture)?;
    let mut document = Yaml::load_from_str(&contents)?;
    for yaml in &mut document {
        let Some(header) = yaml
            .as_mapping_get_mut("then")
            .and_then(|then| then.as_mapping_get_mut("header"))
            .and_then(|header| header.as_sequence_mut())
        else {
            return Ok(());
        };

        header.retain(|header| {
            header.as_mapping_get("name").and_then(|v| v.as_str()) == Some("content-type")
        });
    }

    let mut buf = String::new();
    for yaml in &document {
        YamlEmitter::new(&mut buf).dump(yaml)?;
        buf.push('\n');
    }
    fs::write(fixture, buf.trim_start_matches("---\n"))?;

    Ok(())
}
