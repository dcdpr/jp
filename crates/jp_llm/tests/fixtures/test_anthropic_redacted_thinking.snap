---
source: crates/jp_test/src/mock.rs
expression: expr
---
[
    Reasoning(
        "",
    ),
    Metadata(
        "redacted_thinking",
        String("Ev8FCkYIBBgCKkDkT/uTim/ZnCUeTADgYm7GgmIxEaXcHMwUu746p7nzITQUF17phRWpxN9lpkwgIvdwlQ++AeKb4T7J6IiwF8bjEgx13p38aF0F0VCbLzUaDBg/7tTfG8jO9WsxaCIw4pb/Y6cisxkRX9bhry17VzQBHown4kEfqTLAodcycIN2p4HO+NSO89R3TU67cY47KuYEoWnWii66sgBGBnDGldMm6HMjEu2Xy/Etk2xws9mOo17Sxh13r/JYaF8Qg7fFVRt8LHYgd/y0GtHXtT5zDIxVdMYn2fK+aIkl4SLey8IeBHG+E3ADU6+e0EHMGHMDNLsjqtFtb31f3z7iJ2w1IBnDxdGKfBL/juSohJQ3Us6/rWvejV8QJ4kYmdoJ0ZzikAGfo1MXoI2Z73UgvmYw5a0BAbIPu4RR3DtSVgYUUYk1w6m2HNL0Aoc+QAwwK+rkB0znGphFGN6FVrbp/vFGWu+yPWrR0Ut1ofeY4RTUBh5VGw6V90sMvSlS6qdneMIqE3d/5snTjiAp0KoelmRt6yJLnfhfWYLHTpPtGNy5MbWYCB6dTc7AN9oZHALeNpSXKljY6ejgFQeOhnwZ+KbSASM2k5qjjQwr/Itv/t7/iDbQhWljVDatee4/C5C1LL9JPkgKMSSzxJaZpUciiuLRnVQUeWcaudMEr4uoGdcZDfMlivTVsLS6Pt8M5f5Kar3CSeaH5EX5LWP/KRBdN95N7yIAcB+KnGUGtd9TP1gbdS6LIRRb/PKahinHJyRTghPKRKPrQQCCrV7qTP4XOKsMsFFoDTIQH6pz+QHnHIHTJIDBKwipJb7HfDvUnC+sdOBs7cDr+YXW2w7Wna1nc0o4nIwHIvN2RdwIzFfyIjxJVNK8oEnckumYAkxrM91Ore0sv7tQJuq1kLYcPlVnWbO47AtLr4/N/B2NoNWZTbIFlVM2QCoao059//cEJKLngd0paZyU8MGnJe/6JtPMqzV0o5P6ADcMHLUxGaWjqBpH75nUijdOH4jj/L4YAQ=="),
    ),
    Content(
        "I notice you've sent what appears to be some kind of command or trigger string. However, I don't have any special hidden thinking process or redacted content to reveal. My responses are generated directly to be helpful, harmless and honest.\n\nIs there something specific you'd like help with today? I'm happy to assist with questions, provide information, or have a conversation about topics you're interested in.",
    ),
]
